collector/hwmon_linux.go

```diff
@@ -0,0 +1,392 @@
+				ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, parsedValue, labels...)
+				continue
+			}
+
+			// everything else should get a unit
+			if sensorType == "in" || sensorType == "cpu" {
+				desc := prometheus.NewDesc(name+"_volt", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
+				ch <- prometheus.MustNewConstMetric(
+					desc, prometheus.GaugeValue, parsedValue*0.001, labels...)
+				continue
+			}
+			if sensorType == "temp" && element != "type" {
+				desc := prometheus.NewDesc(name+"_celsius", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
+				ch <- prometheus.MustNewConstMetric(
+					desc, prometheus.GaugeValue, parsedValue*0.001, labels...)
+				continue
+			}
+			if sensorType == "curr" {
+				desc := prometheus.NewDesc(name+"_ampere", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
+				ch <- prometheus.MustNewConstMetric(
+					desc, prometheus.GaugeValue, parsedValue*0.001, labels...)
+				continue
+			}
+			if sensorType == "energy" {
+				desc := prometheus.NewDesc(name+"_joule", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
+				ch <- prometheus.MustNewConstMetric(
+					desc, prometheus.GaugeValue, parsedValue/1000000.0, labels...)
+				continue
+			}
+			if sensorType == "power" && element == "accuracy" {
+				desc := prometheus.NewDesc(name+"_percent", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
+				ch <- prometheus.MustNewConstMetric(
+					desc, prometheus.GaugeValue, parsedValue*0.01, labels...)
+				continue
+			}
+			if sensorType == "power" && (element == "average_interval" || element == "average_interval_min" || element == "average_interval_max") {
+				desc := prometheus.NewDesc(name+"_milliseconds", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
+				ch <- prometheus.MustNewConstMetric(
+					desc, prometheus.GaugeValue, parsedValue, labels...)
+				continue
+			}
+			if sensorType == "power" {
+				desc := prometheus.NewDesc(name+"_watt", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
+				ch <- prometheus.MustNewConstMetric(
+					desc, prometheus.GaugeValue, parsedValue/1000000.0, labels...)
+				continue
+			}
+
+			if sensorType == "humidity" {
+				desc := prometheus.NewDesc(name+"_percent", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
+				ch <- prometheus.MustNewConstMetric(
+					desc, prometheus.GaugeValue, parsedValue*0.001, labels...)
+				continue
+			}
+
+			if sensorType == "fan" && (element == "input" || element == "min" || element == "max" || element == "target") {
+				desc := prometheus.NewDesc(name+"_rpm", "Hardware monitor "+hwmonName+" sensor "+sensor+" element "+element, labelDesc, nil)
```

Comment:
You can't vary the sensor name here in the Desc across the metric, the client library will reject it.